### PR TITLE
Update to 2018 operation

### PR DESCRIPTION
**Pull request by egimbernat**
** Modifications for kops script to support empty S3 bucket
** AlertManager more robust template
** Prometheus operator update to latest version
** Logging (EFK) modified to the last version and based on Kubernetes official example
** Grafana base root_url
** More explanation to readme.md

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -21,12 +21,18 @@ Write your secrets in secrets.yaml, see secrets.yaml.sample.
 The ./kops script supports a few subcommands and a number of arguments:
 
 ### Create a Cluster
-The create subcommand allows you to create a Kubernetes cluster within your AWS account.
+The create subcommand allows you to create a Kubernetes cluster within your AWS account. You must have AWS CLI configured and S3 bucket created.
+
+See [Setting AWS cluster](https://github.com/kubernetes/kops/blob/master/docs/aws.md) for more information.
```
Comment (aknuds1) [inline README.md:26]: I'm not sure what you mean by "Setting AWS cluster" here to be honest, maybe we should use the original title of "Getting started on AWS"?

File: README.md [hunk 2]
```diff
@@ -21,12 +21,18 @@ Write your secrets in secrets.yaml, see secrets.yaml.sample.
...
+
+```
+./kops --aws_profile <preconfigured-aws-profile> create <cluster-name> <cluster-category> <email-for-notifications> <smtp-domain> <base-domain-for-cluster> <s3-bucket-name>
+```
 
 Example:
 
 ```
-./kops --aws_profile example create app staging contact@example.com email-smtp.eu-west-1.amazonaws.com:587 example.com clusters.example.com
+./kops --aws_profile example create my-cluster staging contact@example.com email-smtp.eu-west-1.amazonaws.com:587 example.com my-cluster-bucket
```
Comment (aknuds1) [inline README.md:35]: Does it help to change these example values?

File: README.md [hunk 3]
```diff
@@ -50,4 +56,15 @@ manifests for the rest of the stack. Prometheus Operator will take care of upgra
 components in the cluster.
 
 ### Visualization
-We use Grafana for visualizing monitoring information.
+We use Grafana for visualizing monitoring information, is pre-configured with useful dashboards.
```
Comment (aknuds1) [inline README.md:59]: I think I would drop the "is" here, then it sounds more right.

File: assets/alertmanager-config.yaml [hunk 4]
```diff
@@ -1,21 +1,98 @@
+# This is a sample implementation and should not take it as-is for production
```
Comment (aknuds1) [inline assets/alertmanager-config.yaml:1]: Please change the last part to "and should not be taken as-is for production".

File: kops [hunk 5]
```diff
@@ -176,15 +178,15 @@ client:
...
+  http_auth: {}{}{}
   timeout: 30
   master_only: False
 logging:
   loglevel: INFO
   logfile:
   logformat: default
   blacklist: ['elasticsearch', 'urllib3']
-""".format(logging_secrets['esUsername'], logging_secrets['esPassword'])
+""" .format(curator_secrets['esUsername'], ':' if curator_secrets["authentication"] else None, curator_secrets['esPassword'])
```
Comment (aknuds1) [inline kops:189]: Isn't it cleaner to use `''` instead of `None` here? I'm not even sure what the semantics is here for passing None instead of a string.

Comment (aknuds1) [inline kops:189]: Please use single quotes instead of double quotes for Python strings.

File: kops [hunk 6]
```diff
@@ -208,13 +210,19 @@ def _install_addons(kubecfg_path, args):
             _info('Waiting for cluster to become ready...')
             time.sleep(5)
         else:
+            time.sleep(5)
```
Comment (aknuds1) [inline kops:213]: Why do you add this sleep?

File: kops [hunk 7]
```diff
@@ -208,13 +210,19 @@ def _install_addons(kubecfg_path, args):
...
             break
 
     install_addon('heapster.yaml')
+    time.sleep(5)
```
Comment (aknuds1) [inline kops:217]: Why do you add sleeps between installing each add-on, I have never seen any need for this?

File: kops [hunk 8]
```diff
@@ -255,14 +263,17 @@ def _wait_for_cluster(kubecfg_path):
...
     existing_clusters = _run_kops(
-        ['get', 'clusters'], args, ignore_dry_run=True, get_output=True
-    ).splitlines()
+        ['get', 'clusters'], args, ignore_dry_run=True, get_output=True, tolerate_error=True
+    )
+
+    if existing_clusters is None:
+        return 0
+
+    existing_clusters.splitlines()
```
Comment (aknuds1) [inline kops:276]: I don't think this will work, since `str.splitlines` doesn't mutate, but returns a new string.

File: kops [hunk 9]
```diff
@@ -255,14 +263,17 @@ def _wait_for_cluster(kubecfg_path):
...
+    """Find out if a cluster with the same name exists and if so, asign a number to make it different"""
     re_cluster = re.compile(r'{}-(\d+)'.format(cluster_name), re.I)
     existing_clusters = _run_kops(
-        ['get', 'clusters'], args, ignore_dry_run=True, get_output=True
-    ).splitlines()
+        ['get', 'clusters'], args, ignore_dry_run=True, get_output=True, tolerate_error=True
+    )
+
+    if existing_clusters is None:
+        return 0
```
Comment (aknuds1) [inline kops:274]: Please stick with the old behaviour and start 1 instead of 0.

File: kops [hunk 10]
```diff
@@ -273,13 +284,19 @@ def _cmd_create(args):
...
+def _cmd_create(args):
+    """Invoke kops to create cluster."""
+    cluster_name = '{}.{}'.format(args.category, args.name)
+
+    cluster_number =_check_existing_cluster(args, cluster_name)
+
     # Use three availability zones and one master for each in order to obtain
     # high availability
-    zones = 'eu-central-1a,eu-central-1b,eu-central-1c'
+    zones = 'us-west-2a,us-west-2b,us-west-2c'
```
Comment (aknuds1) [inline kops:299]: Why do you change these zones? From reviewing the code, I think the zones should be configurable by the user though instead of hardwired. I would make these values parameterized in another PR.

File: kops [hunk 11]
```diff
@@ -298,7 +315,7 @@ def _cmd_create(args):
         _run_kops([
             'create', 'cluster', '--zones', zones, '--master-zones', zones,
             '--topology', 'private', '--networking', 'flannel',
-            '--master-size', 'm4.large', '--node-size', 'm4.large',
+            '--master-size', 't2.medium', '--node-size', 't2.medium',
```
Comment (aknuds1) [inline kops:318]: Why do you change these values? Again I think they should be user configurable though instead of hardwired. I would make these values parameterized in another PR.

File: kops [hunk 12]
```diff
@@ -255,14 +263,17 @@ def _wait_for_cluster(kubecfg_path):
...
-def _cmd_create(args):
-    """Invoke kops to create cluster."""
-    cluster_name = '{}.{}'.format(args.category, args.name)
+def _check_existing_cluster(args, cluster_name):
+    """Find out if a cluster with the same name exists and if so, asign a number to make it different"""
     re_cluster = re.compile(r'{}-(\d+)'.format(cluster_name), re.I)
     existing_clusters = _run_kops(
-        ['get', 'clusters'], args, ignore_dry_run=True, get_output=True
-    ).splitlines()
+        ['get', 'clusters'], args, ignore_dry_run=True, get_output=True, tolerate_error=True
```
Comment (aknuds1) [inline kops:270]: Will this fail if the kops bucket doesn't already exist and `tolerate_error` is false?

File: kops [hunk 13]
```diff
@@ -273,13 +284,19 @@ def _cmd_create(args):
...
-            break
+            return cluster_number
     else:
         _error('Couldn\'t determine a unique cluster name')
 
+def _cmd_create(args):
+    """Invoke kops to create cluster."""
+    cluster_name = '{}.{}'.format(args.category, args.name)
+
+    cluster_number =_check_existing_cluster(args, cluster_name)
```
Comment (aknuds1) [inline kops:295]: You lack a space after `=`.

File: kops [hunk 14]
```diff
@@ -420,10 +437,10 @@ def _run_kops(
                 command, env={**os.environ, **merged_env}
             )
             proc.wait()
-        if proc.returncode != 0:
+        if proc.returncode != 0 and not tolerate_error:
             _error('kops failed with code {}'.format(proc.returncode))
         if get_output:
-            return stdout
+            return stdout if proc.returncode == 0 else None # If None is returned, an error ocurred and 'tolerate_error' has been setted
```
Comment (aknuds1) [inline kops:443]: Please put the comment on the line above as line lengths should be <= 80.

File: secrets.yaml.sample [hunk 15]
```diff
@@ -4,3 +4,8 @@ alertmanager:
 logging:
   esS3AccessKey: example
   esS3SecretKey: example
+# If you are using X-Pack of elastic you must change below values
```
Comment (aknuds1) [inline secrets.yaml.sample:7]: Please change to 'If you are using Elastic X-Pack you must change the below values'

**Review by aknuds1 — CHANGES_REQUESTED** [review]

File: kops (hunk 7 above)
Comment (egimbernat) [reply kops:217]: I added it because for some reason kops cli where not ready for the next instruction and makes Python kops script to crash. Probably because of bandwidth where I live?

File: kops (hunk 10 above)
Comment (egimbernat) [reply kops:299]: I agree, the zones and machine size should be configurable. I first wanted to be sure that my PR was on the desired direction before making more changes.

File: kops (hunk 7 above)
Comment (aknuds1) [reply kops:217]: @egimbernat It can't be the `kops` command that fails since `install_addon` calls `kubectl apply`, i.e. it would have to be `kubectl` that would fail. I don't see why `kubectl` would fail to issue requests in rapid succession, could you try asking in the Kubernetes Slack for example to find out why this might happen?

**Comment by kojiwakayama** [comment]
Hey @egimbernat, thanks for the PR! Would it be possible to split up the PR into smaller ones so we can get the changes in quicker? :)

**Comment by aknuds1** [comment]
@egimbernat I made the change last night btw to parameterize AZs and node sizes, so you don't have to worry about that in your PR. I would suggest to, as @kojiwakayama mentions, pare the PR down to updating the add-ons and not changing the kops script. If you want to change the kops script, please issue another PR :)

File: kops (hunk 7 above)
Comment (kojiwakayama) [reply kops:217]: we updated the cluster validation functionality which might resolve the issue. https://github.com/coder-society/k8s-aws/pull/10

**Comment by egimbernat** [comment]
Hello @kojiwakayama @aknuds1 , sure. I'll split it in little PR so can be integrated more quickly. I close this PR for then create the others.
